consensus/src/chained_bft/liveness/leader_reputation.rs

```diff
@@ -6,21 +6,90 @@ use consensus_types::{
     block::Block,
     common::{Author, Round},
 };
-use libra_types::block_metadata::BlockMetadata;
+use libra_logger::prelude::*;
+use libra_types::block_metadata::{new_block_event_key, NewBlockEvent};
+use libradb::LibraDBTrait;
 use serde::export::PhantomData;
-use std::{cmp::Ordering, collections::HashSet, sync::Arc};
+use std::{
+    cmp::Ordering,
+    collections::HashSet,
+    sync::{Arc, Mutex},
+};
 
 /// Interface to query committed BlockMetadata.
-pub trait MetadataBackend {
-    /// Return a window_size contiguous BlockMetadata window in which last one is at target_round or
+pub trait MetadataBackend: Send + Sync {
+    /// Return a contiguous BlockMetadata window in which last one is at target_round or
     /// latest committed, return all previous one if not enough.
-    fn get_block_metadata(&self, window_size: u64, target_round: Round) -> &[BlockMetadata];
+    fn get_block_metadata(&self, target_round: Round) -> Vec<NewBlockEvent>;
+}
+
+pub struct LibraDBBackend {
+    window_size: usize,
+    libra_db: Arc<dyn LibraDBTrait>,
+    window: Mutex<Vec<(u64, NewBlockEvent)>>,
+}
+
+impl LibraDBBackend {
+    pub fn new(window_size: usize, libra_db: Arc<dyn LibraDBTrait>) -> Self {
+        Self {
+            window_size,
+            libra_db,
+            window: Mutex::new(vec![]),
+        }
+    }
+
+    fn refresh_window(&self, target_round: Round) -> anyhow::Result<()> {
+        // assumes target round is not too far from latest commit
+        let buffer = 10;
+        let events = self.libra_db.get_events(
+            &new_block_event_key(),
+            u64::max_value(),
+            self.window_size as u64 + buffer,
+        )?;
+        let mut result = vec![];
+        for (v, e) in events {
+            let e = lcs::from_bytes::<NewBlockEvent>(e.event_data())?;
+            if e.round() <= target_round && result.len() < self.window_size {
+                result.push((v, e));
+            }
+        }
+        *self.window.lock().unwrap() = result;
+        Ok(())
+    }
+}
+
+impl MetadataBackend for LibraDBBackend {
+    // assume the target_round only increases
+    fn get_block_metadata(&self, target_round: Round) -> Vec<NewBlockEvent> {
+        let (known_version, known_round) = self
+            .window
+            .lock()
+            .unwrap()
+            .first()
+            .map(|(v, e)| (*v, e.round()))
+            .unwrap_or((0, 0));
+        if !(known_round == target_round
+            || known_version == self.libra_db.get_latest_version().unwrap_or(0))
```

Comment:
maybe assign the latest_version outside the if statement so that this is a little cleaner to read?

Reply:
I was trying to use short circuit but later added the !